chapters/part3/document-ranking.md

```diff
@@ -348,7 +348,91 @@ Next, we need to evaluate the model's performance. There are several metrics com
 | elmo           | 0.4 (2)           | 0.55 (1)               |
 | grover         | 0.1 (3)           | 0.05 (3)               |
 
-The correct option is the cookie monster, but our model only ranked it second. It shouldn't get a complete zero for that, but shouldn't be fully rewarded either. The reciprocal rank would give this a `1/2` for ranking the best option 2nd. More generally, the reciprocal rank score is `1` divided by the position your model ranked the correct option. If it had correctly ranked cookie monster first, its score would be `1/1 = 1`. As the name suggests, we take the mean of all these reciprocal rank scores across our dataset to find the MRR.
+The correct option is the cookie monster, but our model only ranked it second. It shouldn't get a complete zero for that, but shouldn't be fully rewarded either. The reciprocal rank would give this a `1/2` for ranking the best option 2nd. More generally, the reciprocal rank score is `1` divided by the position your model ranked the correct option. If it had correctly ranked cookie monster first, its score would be `1/1 = 1`. As the name suggests, we take the mean of all these reciprocal rank scores across our dataset to find the MRR. This isn't too hard to implement in Pytorch:
 
+```python
+def first_nonzero(t):
+    idx = torch.arange(t.size(1), 0, -1).type_as(t)
+    indices = torch.argmax(t * idx, 1, keepdim=True)
+    return indices
+
+def mrr(y_pred, y_true, mask):
+    y_pred = y_pred.masked_fill(~mask, -1)
+    y_true = y_true.ge(y_true.max(dim=-1, keepdim=True).values).long()
+
+    _, idx = y_pred.sort(descending=True, dim=-1)
+    gold = torch.arange(y_true.size(-1)).unsqueeze(0).type_as(y_pred)
+    
+    ordered_truth = y_true.gather(1, idx)
+    _mrr = (ordered_truth / (gold + 1)) * mask
+    
+    return _mrr.gather(1, first_nonzero(ordered_truth)).mean()
+
+y_pred = torch.tensor([[0.25, 0.55, 0.05]])
+y_true = torch.tensor([[0.65, 0.4, 0.1]])
+mask = torch.ones_like(yt).bool()
+
+mrr(y_pred, y_true, mask)
+> tensor(0.5000)
+```
+
+We binarize the labels by setting the highest-scored document to have label `1`, and the rest `0` (note that there are other ways you could do this). 
```

Comment:
Here we get some guidance.  I guess I should have read this before trying to understand the code...  But a reader will probably do what I did, because the code shows up first.